ql/CMakeLists.txt

```diff
@@ -2332,6 +2332,10 @@ target_compile_definitions(ql_library PRIVATE
 target_compile_options(ql_library PRIVATE
     ${OpenMP_CXX_FLAGS})
 
+if(NOT "${QL_EXTRA_LINK_LIBRARIES}" STREQUAL "")
+    target_link_libraries(ql_library PUBLIC ${QL_EXTRA_LINK_LIBRARIES})
```

Comment:
Is this needed because you can't use `target_link_libraries(ql_library PUBLIC ...)` inside of the subdirectory that you add?

Reply:
This is technically correct, but the `QL_EXTRA_LINK_LIBRARIES` is more flexible. For example if QuantLib itself would be used as a subproject in something else, this variable could be used without the external subdirectory feature above.

Reply:
I believe that @pcaspers uses QuantLib as a subproject (i.e. git submodule), so it might be interesting to hear how he specifies additional libraries to be linked, if any.

Reply:
If you mean https://github.com/OpenSourceRisk/Engine, the submodule is built normally and linked against the other parts. Nothing links against QuantLib itself. Also, the submodule is not the official QuantLib version from this repo, but a modified fork.

However, allowing the extension flag above would give submodules a great deal of flexibility, adding flags and settings to QuantLib itself without using a fork.

Reply:
Yes, different use case.  In ORE, QuantLib is a submodule and is linked with others in an external executable.  The use case that this PR covers is, for example, replacing double with some other type which needs to link a library.  In this case, parts of QuantLib itself (say, the test suite) need to link the latter.

Reply:
Yes I think this "extension point" here is rather unusual. A library would normally provide template functions where several parameter types make sense. 

The "Real" typedef is more a relict from very long ago to allow for builds with float or double (I believe)? It should really be double in my opinion.

And remember plugging in an AD type for "Real" across the whole library is not for free - we (and others) played with that as well in the past using different AD libs and the results were at best "mixed". A selective approach is more effort, but in my opinion there is no way around that really.